content/docs/command-reference/params/index.md

```diff
@@ -28,6 +28,9 @@ YAML 1.2, JSON, TOML, or [Python](#examples-python-parameters-file) _parameters
 file_. Multiple parameter dependencies can be specified from one or more
 parameters files.
 
+💡 Parameters can also be used as variables in `dvc.yaml` itself. See the
+[Parametrizes stages](#example-parametrized-stages-in-dvcyaml) example below.
```

Comment:
I find this feature pretty confusing. Params were originally meant as values for your code, which you can track granularly with DVC. Now we're also enabling their use as variables inside dvc.yaml itself, where they can also be define as params for the code. You can even do both things at once. Meta and kind of confusing.

In fact I'm avoiding calling this "parametrization" (of stages/dvc.yaml) in docs to lessen the confusion.

This has probably been discussed internally so for now I'm just commenting here cc @shcheklein @skshetry thoughts?

For now I'll try to write the docs encouraging the use of `vars` sections mainly, and just mentioning this as an alternative.

Reply:
p.s. an obvious proposal is to move this functionality to a file named `vars.yaml` (by default, configurable) to clearly separate code params from dvc.yaml vars.

Reply:
> You can even do both things at once (with the same param)

This is one of the problems that it tries to solve. Is it a bad thing?

> In fact I'm avoiding calling this "parametrization" (of stages/dvc.yaml) in docs to lessen the confusion.

I like to call the values "variables" (though they are static 😕), but the stage definition does have parameters to be passed (like in functions), so parametrization?

Another term could be "templating".

Reply:
Not sure about the naming - vars, etc.

>You can even do both things at once (with the same param)

I also don't see a problem with this. Consider your pipeline definition part of the project/your code. That's it, you parametrize all things this way. `Parameters can also be used as variables` - this phrase sounds a bit confusing indeed though. I would need to read more to recommend something better.

Reply:
> This is one of the problems that it tries to solve.
> Consider your pipeline definition part of the project/your code.

OK so you have `var: 'value'` in params.yaml and use that value in your code.py, but you also use it as part of the definition of the stage that execs code.py? Resulting in something like

```yaml
stages:
  my-thing:
    cmd: python code.py # reads params.yaml, uses var
    params:
      - var # dvc status/repro will check this from params.yaml 
    outs:
      - ${var}.txt
```

I'm just not seeing the logic for this use case. But maybe an example would help me understand and explain this @skshetry 🙂

Reply:
In any casem that's not my main concern. That would be that DVC reads params.yaml twice, for separate reasons:

1. to substitute it's values in the dvc.yaml definition itself; and
2. (status, repro) to determine if the stage is outdated

But it never passes the values to code.py BTW, which has to read it a third time.

It seems to me like params files are being overloaded with features which may not be easy to grasp by most users.

Reply:
` # reads params.yaml, uses var` -  why does it have to use it?

Reply:
> OK so you have `var: 'value'` in params.yaml and use that value in your code.py, but you also use it as part of the definition of the stage that execs code.py?

The main use case of this is to pass values in `cmd` itself (this way it will be tracked in `params diff`, so we don't need to add it in `params` section).

```yaml
stages:
 my-thing:
   cmd: python code.py --thresh ${thresh}
```

Better to start with some simple examples.

Reply:
> why does it have to use it?

@shcheklein it doesn't have to but it could and that's the scenario I'm presenting.

> Main use: pass values in cmd itself (this way it will be tracked in params diff

Hmm... interesting. I didn't know `params` subcommands would track values if they're not in the `params` field. Let me try it out ⌛

Reply:
Oh I see what you mean: everything in params.yaml is used by `params diff` anyway (whether or not is used at all in dvc.yaml). OK, that makes sense... I still think that the feature is overloaded in a possibly confusing way but I'll leave it up to you guys.

As for the case of using the same param both in the stage def (in dvc.yaml) AND in the stage code, I'll just regard that as a side-effect which we don't need to emphasize (maybe not even mention directly) in docs for now.

> Consider your pipeline definition part of the project/your code. That's it, you parametrize all things this way.

I'll try to adopt this perspective as well to keep and generalize the "parametrization" concept/term...

Reply:
> As for the case of using the same param both in the stage def (in dvc.yaml) AND in the stage code, I'll just regard that as a side-effect which we don't need to emphasize (maybe not even mention directly) in docs for now...

Hmm, I still don't quite understand the concern. Why should we limit it any way? Params "parametrize" _all_ the code. `dvc.yaml` is part of your code. Thus params can be used to parametrize it. It's up to users to use params anywhere they want to my mind.

Reply:
> Not sure about the naming - vars, etc.
> `Parameters can also be used as variables` - this phrase sounds a bit confusing indeed

@shcheklein I've gone for "parameteriz(e|ation)" and "templat(e|ing)" for now. I'm avoiding "variables" and also opened iterative/dvc/issues/5164 about the naming of the currently `vars` section cc @skshetry.

> Why should we limit it any way?

I'm not limiting it. I'm just not planning to complicate the docs explaining this particular case (using the same param in dvc.yaml as in the code it executes) for now. Maybe if this was a cmd ref or when/if we do extract this part of the guide into it's own (true) guide, we can then add examples and cover that specifically.

Reply:
> Oh I see what you mean @skshetry: everything in params.yaml is used by params diff anyway

So, any files specified in the `params` section and `params.yaml` by default are shown in the `params diff`. This includes every entry, not just the specified "param" keys. That's just how `params diff` work (which might change in the future). 🤷🏽 

In addition to that, parametrized variables are also displayed on `params diff`. This will be clear if you specify vars as follows for example (and not include them on `params` section):
```
vars:
  - test_params.yaml
```

Reply:
@jorgeorpinel, please don't confuse `params` with `vars`. They are different things and for different purposes. Trying to mix them will only make it more confusing.

`dvc.yaml` always defaults to `params.yaml` for `vars` and `params`. The defaults might be making you see them as same/similar thing, but they are not. As an example:

```yaml
vars:
  - test_params.yaml
stages:
  stage1:
    cmd: "${script} --thresh ${thresh}"
    params:
      - script
      - thresh
```

Here, the `script` and `thresh` are from `params.yaml` in `params` section whereas `script` and `thresh` in the `cmd` might be coming from `test_params.yaml` or `params.yaml` (default). This is because `vars` is actually recursively merged together.

Reply:
`params` section defaults to `params.yaml` is also incorrect though. It always defaults to the `params.yaml` of the `wdir` (which defaults to `dvc.yaml`'s directory).

This is the same for `vars` as well in the `stages` though. So, that means `wdir` is always resolved from global `vars` section as local `vars` section cannot be loaded before `wdir`.

Reply:
> In addition to that, parametrized variables are also displayed on params diff...

Good to know. A bunch of notes/clarifications are still missing from here (last check box left unmarked in the description of this PR). For now let's focus on getting the main cases clearly explained.

> params and vars are different things and for different purposes. Trying to mix them will only make it more confusing

I think they overlap a lot i.e. they have the same purpose in the end, which is "parameterizing" dvc.yaml. They are already mixed in the way this feature was designed so if that is confusing I think we should discuss separately (e.g. in https://github.com/iterative/dvc/issues/5164). For now I'm following Ivan's suggestion to just consider everything "parameterization" which I do agree now is the simplest path for this iteration of docs.

> params section ... always defaults to the params.yaml of the wdir
> the same for vars as well in the stages... wdir is always resolved from global vars section

Yep, again one of the many notes still missing, but thanks for the reminder.